darwin fips builds

a-palchikov opened the pull request:
This PR serves the purposes of testing the build pipeline.

Requires https://github.com/gravitational/teleport.e/pull/227 and https://github.com/gravitational/godl/pull/1.

webvictim commented on .drone.yml:
```diff
@@ -904,6 +904,143 @@ steps:
+      GOPATH: /tmp/push-build-darwin-amd64-fips/go
+      GOCACHE: /tmp/push-build-darwin-amd64-fips/go/cache
+      # use custom Go compiler
+      GO_BINARY: /tmp/push-build-darwin-amd64-fips/go/src/github.com/gravitational/godl/build/dev.boringcrypto.go1.15
```
Find a way to set this using the file containing the value `BORINGCRYPTO_BRANCH` so we don't need this repetitive `dev.boringcrypto.go1.15` string set everywhere.

webvictim commented on .drone.yml:
```diff
@@ -2937,6 +3074,287 @@ steps:
+      WORKSPACE_DIR: /tmp/build-darwin-amd64-fips
+      TELEPORT_DIR: /tmp/build-darwin-amd64-fips/go/src/github.com/gravitational/teleport
+      GODL_DIR: /tmp/build-darwin-amd64-fips/go/src/github.com/gravitational/godl
+      GODL_COMMIT: dev.boringcrypto.go1.15
```
As above, set this once per pipeline and use it throughout the other steps.

webvictim commented on .drone.yml:
```diff
@@ -2937,6 +3074,287 @@ steps:
+      - set -u
+      - cd $WORKSPACE_DIR/go/src/github.com/gravitational/teleport
+      - export VERSION=$(cat $WORKSPACE_DIR/go/.version.txt)
+      - make pkg OS=$OS ARCH=$ARCH
```
This calls the `build-package.sh` script which is going to need to be made FIPS-aware for Mac. You should add it to the command:

```suggestion
      - make pkg OS=$OS ARCH=$ARCH FIPS=yes
```

webvictim reviewed (commented):
I notice that there's no `build-darwin-amd64-fips-pkg-tsh` target - we should copy that over too and make the appropriate changes.

webvictim commented on .drone.yml:
```diff
@@ -904,6 +904,143 @@ steps:
+    commands:
+      - set -u
+      - cd $WORKSPACE_DIR/go/src/github.com/gravitational/teleport
+      - make clean release OS=$OS ARCH=$ARCH FIPS=$FIPS GO_BINARY=$GO_BINARY
```
We don't provide/create FIPS builds for OSS, so it might be better to run something like this:

```suggestion
      - make -C e clean release VERSION=$VERSION OS=$OS ARCH=$ARCH FIPS=$FIPS GO_BINARY=$GO_BINARY
```

You'd need to set the `VERSION` variable using similar logic to what we do here: https://github.com/gravitational/teleport/blob/master/.drone.yml#L1748-L1769

webvictim commented on .drone.yml:
```diff
@@ -904,6 +904,150 @@ steps:
+      AWS_SECRET_ACCESS_KEY:
+        from_secret: AWS_SECRET_ACCESS_KEY
+      AWS_REGION: us-west-2
+      BORINGCRYPTO_BRANCH: dev.boringcrypto.go1.15
```
Should be set using `export BORINGCRYPTO_BRANCH=$(cat $WORKSPACE_DIR/go/.bc_commit.txt)` in commands as above and removed here.

webvictim commented on .drone.yml:
```diff
@@ -904,6 +904,150 @@ steps:
+      GODL_DIR: /tmp/push-build-darwin-amd64-fips/go/src/github.com/gravitational/godl
+    commands:
+      - set -u
+      - mkdir -p $HOME/sdk/dev.boringcrypto.go1.15
+      - cd $HOME/sdk/dev.boringcrypto.go1.15
```
Set using `$BORINGCRYPTO_BRANCH` rather than hardcoding.

webvictim commented on .drone.yml:
```diff
@@ -2937,6 +3081,294 @@ steps:
+    commands:
+      - set -u
+      - export BORINGCRYPTO_BRANCH=$(cat $WORKSPACE_DIR/go/.bc_commit.txt)
+      - mkdir -p $HOME/sdk/dev.boringcrypto.go1.15
+      - cd $HOME/sdk/dev.boringcrypto.go1.15
```
Get these from `$BORINGCRYPTO_BRANCH` rather than hardcoding

webvictim commented on .drone.yml:
```diff
@@ -2937,6 +3081,294 @@ steps:
+      - cd $WORKSPACE_DIR/go/src/github.com/gravitational/teleport
+      - export VERSION=$(cat $WORKSPACE_DIR/go/.version.txt)
+      - export BORINGCRYPTO_BRANCH=$(cat $WORKSPACE_DIR/go/.bc_commit.txt)
+      - make -C e clean release VERSION=$VERSION OS=$OS ARCH=$ARCH FIPS=$FIPS GO_BINARY="$GODL_DIR/build/$BORINGCRYPTO_BRANCH"
```
I think `GITTAG` needs to be set here

```suggestion
      - make -C e clean release GITTAG=v$VERSION VERSION=$VERSION OS=$OS ARCH=$ARCH FIPS=$FIPS GO_BINARY="$GODL_DIR/build/$BORINGCRYPTO_BRANCH"
```

webvictim commented on .drone.yml:
```diff
@@ -2937,6 +3081,294 @@ steps:
+      - set -u
+      - cd $WORKSPACE_DIR/go/src/github.com/gravitational/teleport
+      # copy release archives to artifact directory
+      - cp teleport*.tar.gz $WORKSPACE_DIR/go/artifacts
```
This step isn't needed as we don't build OSS artifacts for FIPS.

```suggestion
```

webvictim reviewed (commented):
Looking better, few more nits.

You'll also need to modify things like https://github.com/gravitational/teleport/blob/master/build.assets/build-package.sh#L108-L110 to add support for FIPS packages on Mac, as it currently ignores the `RUNTIME` parameter. 

We should also change the identifier (https://github.com/gravitational/teleport/blob/master/build.assets/build-package.sh#L232) for FIPS builds to distinguish them from the regular `tsh` builds (assuming that we need to provide these)

a-palchikov committed:
Switch to go1.16. Use embed package to embed webassets instead of ad-hoc attaching to binary

a-palchikov committed:
Switch to go1.16. Use embed package to embed webassets instead of ad-hoc attaching to binary

russjones commented:
@webvictim @a-palchikov We're just waiting on Go 1.16.2 BoringCrypto at this point right?

a-palchikov commented:
> @webvictim @a-palchikov We're just waiting on Go 1.16.2 BoringCrypto at this point right?

I believe so, yes.

a-palchikov commented:
go1.16.2 has been [merged](https://go-review.googlesource.com/c/go/+/303952) into boringcrypto.go1.16

webvictim commented:
Awesome, once it appears in https://go.googlesource.com/go/+/dev.boringcrypto/misc/boring/RELEASES I guess we can use it?

a-palchikov commented:
> Awesome, once it appears in https://go.googlesource.com/go/+/dev.boringcrypto/misc/boring/RELEASES I guess we can use it?

It's been added the same day I'm guessing. I'll run a couple of tests and let you know when it's good to merge,

a-palchikov committed:
Fix pipeline duplicate step error

a-palchikov committed:
Resolve duplicate pipeline step name error.
Explicitly define platform for 'exec' pipelines.
Remove the uid/gid environment from 'exec' pipelines as redundant.

a-palchikov committed:
Set proper dependencies when building darwin package fips pipelines. Use enterprise build directory for tsh

webvictim commented on dronegen/darwin_pipelines.go:
```diff
@@ -0,0 +1,532 @@
+	"path/filepath"
+)
+
+const boringCryptoBranch = "dev.boringcrypto.go1.16"
```
I would be tempted to put this in common.go next to the `goRuntime` constant so it's obvious that both need to be updated for any future Go version changes.

webvictim reviewed (approved):
LGTM I think

a-palchikov committed:
Address review comments

russjones reviewed (approved):
Bot.

a-palchikov committed:
Bump e to tip
